rmw_connext_dynamic_cpp/src/functions.cpp

```diff
@@ -1263,15 +1266,38 @@ rmw_take(const rmw_subscription_t * subscription, void * ros_message, bool * tak
     return RMW_RET_ERROR;
   }
 
-  bool success = _take(&dynamic_data_sequence[0], ros_message, members);
-  if (!success) {
-    rmw_set_error_string("failed to take sample");
-    return RMW_RET_ERROR;
+  bool success = true;
+  bool ignore_sample = false;
+  if (subscriber_info->ignore_local_publications) {
+    // compare the lower 12 octets of the guids from the sender and this receiver
+    // if they are equal the sample has been sent from this process and should be ignored
+    DDS_SampleInfo & sample_info = sample_infos[0];
+    DDS_GUID_t sender_guid = sample_info.original_publication_virtual_guid;
+    DDS_InstanceHandle_t receiver_instance_handle = dynamic_reader->get_instance_handle();
+    ignore_sample = true;
+    for (uint i = 0; i < 12; ++i) {
+      DDS_Octet * sender_element = &(sender_guid.value[i]);
+      DDS_Octet * receiver_element = &(((DDS_Octet *)&receiver_instance_handle)[i]);
+      if (*sender_element != *receiver_element) {
+        ignore_sample = false;
+        break;
+      }
+    }
+  }
+
+  if (!ignore_sample) {
```

Comment:
If `ignore_sample` is `true` and `success` is `true` the code must set `taken` to `false`. Currently it is not a problem since the only code using the API initializes `taken` with `false` before calling but shouldn't need to.